### PR TITLE
nixos/libinput: Allow 'custom' acceleration profile

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

Expand the allowed enum values for the `accelProfile` option of `services/hardware/libinput ` to allow for `"custom"`, which is [another possible value](https://wayland.freedesktop.org/libinput/doc/latest/pointer-acceleration.html#pointer-acceleration-profiles) in the newer versions of the library. I have tested to make sure the option gets applied correctly.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by pluiedev — APPROVED** [review]
Looks good, but is there any way for the user to define what the acceleration function *actually* is in Nix?

**Comment by VSinerva** [comment]
> Looks good, but is there any way for the user to define what the acceleration function _actually_ is in Nix?

This can be done with `services.xserver.libinput.(mouse/touchpad).additionalOptions` by using the relevan libinput options like `Option "AccelStepMotion"` and `Option "AccelPointsMotion"`. The problem was that the `AccelProfile` enum is forced to *always* have a value, which couldn't be overridden via `additionalOptions`. So you could define an acceleration function, but couldn't set libinput to use the `custom` profile.

**Comment by pluiedev** [comment]
> > Looks good, but is there any way for the user to define what the acceleration function _actually_ is in Nix?
> 
> 
> 
> This can be done with `services.xserver.libinput.(mouse/touchpad).additionalOptions` by using the relevan libinput options like `Option "AccelStepMotion"` and `Option "AccelPointsMotion"`. The problem was that the `AccelProfile` enum is forced to *always* have a value, which couldn't be overridden via `additionalOptions`. So you could define an acceleration function, but couldn't set libinput to use the `custom` profile.

Hm, I see! I would then suggest to add those options (step motion and points motion) to the module too if feasible, and mention those options in the custom profile's description. It should also be possible then to set the profile to custom by default, if those options were set. 

Worst case, just mention those raw options and tell people to add them to their `additionalOptions` if they want the custom profile to do anything.

**Comment by VSinerva** [comment]
> > > Looks good, but is there any way for the user to define what the acceleration function _actually_ is in Nix?
> > 
> > 
> > This can be done with `services.xserver.libinput.(mouse/touchpad).additionalOptions` by using the relevan libinput options like `Option "AccelStepMotion"` and `Option "AccelPointsMotion"`. The problem was that the `AccelProfile` enum is forced to _always_ have a value, which couldn't be overridden via `additionalOptions`. So you could define an acceleration function, but couldn't set libinput to use the `custom` profile.
> 
> Hm, I see! I would then suggest to add those options (step motion and points motion) to the module too if feasible, and mention those options in the custom profile's description. It should also be possible then to set the profile to custom by default, if those options were set.
> 
> Worst case, just mention those raw options and tell people to add them to their `additionalOptions` if they want the custom profile to do anything.

I added all the options relevant to custom accel functions. I held off on making `accelProfile` automatically change to "custom", because the only way I know of doing that would be to change the default value to some conditional mess with all six options being checcked for a non-null values

File: nixos/modules/services/hardware/libinput.nix
Comment (pluiedev) [inline nixos/modules/services/hardware/libinput.nix:50]:
```suggestion
        type = types.nullOr (types.listOf types.number)
```
Ditto for below

File: nixos/modules/services/hardware/libinput.nix
Comment (pluiedev) [inline nixos/modules/services/hardware/libinput.nix:81]:
```suggestion
        type = types.nullOr types.number;
```
Ditto for below

**Review by pluiedev — CHANGES_REQUESTED** [review]
Mmmmmm... I personally tend to model this with a `types.attrTag` that allows options to be specified that are only relevant for the specific enum type, like this:
```nix
{
  custom.stepMotion = [ ... ];
}
# or
{
  adaptive.pointsMotion = [ ... ];
}
```

However I perfectly understand if this couldn't be done due to backwards compatibility reasons, or it's simply too complicated.

File: nixos/modules/services/hardware/libinput.nix
Comment (VSinerva) [reply nixos/modules/services/hardware/libinput.nix:50]: I can see the reason for changing the types, and had considered it. I prioritized consistency with the pre-existing options, as there are already very similar options in the same module that are typed as `str`. Would it be better to ignore them?

**Comment by VSinerva** [comment]
> Mmmmmm... I personally tend to model this with a `types.attrTag` that allows options to be specified that are only relevant for the specific enum type, like this:
> 
> ```nix
> {
>   custom.stepMotion = [ ... ];
> }
> # or
> {
>   adaptive.pointsMotion = [ ... ];
> }
> ```
> 
> However I perfectly understand if this couldn't be done due to backwards compatibility reasons, or it's simply too complicated.

I think I understand the basic point you're making, but I have no idea how to implement it in the Nix language, and my Googling is coming up short :/

If this is something that's important (and you're able to help a newbie through it) I won't object, but otherwise I'd be inclined to leave it.

**Comment by pluiedev** [comment]
TBH I wouldn't worry too much about it. It's going to be a pain to try to migrate all the old settings to the new format and it's really just not worth it — I would totally try to do it myself if it's a new module though

**Comment by VSinerva** [comment]
Is there something more that needs to be done on my part to get this merged?

No rush ofc, just want to make sure I'm not unknowingly holding anything up :)

**Review by pluiedev — APPROVED** [review]
LGTM

**Comment by VSinerva** [comment]
It's been a while and I noticed that this module is not covered by anyone via `CODEOWNERS`, and modules don't have a maintainers list built in, unlike packages. Best I can find is the `freedesktop` team.

@jtojnar Are you able to merge this PR? Sorry if this ping is erroneous! (First time contributing to NixOS)

**Comment by SuperSandro2000** [comment]
Someone wants to take a look at #308777?

**Pull request merged**
